src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/CodeGeneration/CodeWriter.IndentCache.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Microsoft.AspNetCore.Razor.Language.CodeGeneration;
+
+public sealed partial class CodeWriter
+{
+    internal static class IndentCache
+    {
+        internal const int MaxTabCount = 64;
+        internal const int MaxSpaceCount = 128;
+
+        private static readonly ReadOnlyMemory<char> s_tabs = new string('\t', MaxTabCount).AsMemory();
+        private static readonly ReadOnlyMemory<char> s_spaces = new string(' ', MaxSpaceCount).AsMemory();
+
+        public static ReadOnlyMemory<char> GetIndentString(int size, bool useTabs, int tabSize)
+        {
+            if (!useTabs)
+            {
+                return SliceOrCreate(size, s_spaces);
+            }
+
+            var tabCount = size / tabSize;
+            var spaceCount = size % tabSize;
+
+            if (spaceCount == 0)
+            {
+                return SliceOrCreate(tabCount, s_tabs);
+            }
+
+            return string.Create(length: tabCount + spaceCount, (tabCount, spaceCount), static (destination, state) =>
+            {
+                var (tabCount, spaceCount) = state;
+
+                var tabs = SliceOrCreate(tabCount, s_tabs);
+                var spaces = SliceOrCreate(spaceCount, s_spaces);
```

Comment:
However unlikely, it doesn't feel right to potentially allocate strings inside of a call to `string.Create(...)` since that's intended for allocation-free string creation. On the one hand, it does read as much simpler but that makes the potential allocation even more subtle. Personally, I probably wouldn't have cut this particular corner, but I do understand your thinking.